pkg/shifter/shifter.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/roppenlabs/silent-assassin/pkg/killer"
+	"github.com/roppenlabs/silent-assassin/pkg/logger"
+	"github.com/roppenlabs/silent-assassin/pkg/notifier"
+	container "google.golang.org/api/container/v1"
+	v1 "k8s.io/api/core/v1"
+)
+
+const timeLayout string = "15:04"
+
+//Suggest a better name
+type npShiftConf struct {
+	destNodePool       string
+	sourceMinNodeCount int64
+}
+
+type ShifterService struct {
+	cp                 config.IProvider
+	logger             logger.IZapLogger
+	kubeClient         k8s.IKubernetesClient
+	gcloudClient       gcloud.IGCloudClient
+	killer             killer.KillerService
+	notifier           notifier.INotifierClient
+	nodePoolMap        map[string]npShiftConf
+	whiteListIntervals []wlInterval
+}
+
+func NewShifterService(cp config.IProvider, zl logger.IZapLogger, kc k8s.IKubernetesClient, gc gcloud.IGCloudClient, nf notifier.INotifierClient, kl killer.KillerService) ShifterService {
+	return ShifterService{
+		cp:           cp,
+		logger:       zl,
+		kubeClient:   kc,
+		gcloudClient: gc,
+		notifier:     nf,
+		killer:       kl,
+	}
+}
+
+func (ss ShifterService) Start(ctx context.Context, wg *sync.WaitGroup) {
+	ss.logger.Info(fmt.Sprintf("Starting Shifter Loop - Poll Interval: %d", ss.cp.GetInt(config.ShifterPollIntervalMs)))
+	ss.initWhitelist()
+
+	for {
+		select {
+		case <-ctx.Done():
+			ss.logger.Info("Shutting down Shifter service")
+			wg.Done()
+			return
+		default:
+			// Check if current time is in bettween whiteListIntervals. If yes, run ss.shift().
+			t := time.Now().UTC()
+			hour := fmt.Sprintf("%02d", t.Hour())
+			mins := fmt.Sprintf("%02d", t.Minute())
+			fmtTime := fmt.Sprintf("%s:%s", hour, mins)
+			now, err := time.Parse(timeLayout, fmtTime)
+
+			if err != nil {
+				panic(err)
+			}
+
+			for _, interval := range ss.whiteListIntervals {
+				if timeWithinWLIntervalCheck(interval.start, interval.end, now) {
+					ss.shift()
+				}
+			}
+			time.Sleep(time.Millisecond * time.Duration(ss.cp.GetInt(config.ShifterPollIntervalMs)))
+		}
+	}
+}
+
+//getNodePoolMap finds out fallback on-demand nodePools and their respective preemptible node-pools.
+func (ss *ShifterService) getNodePoolMap() error {
+
+	nps, err := ss.gcloudClient.ListNodePools()
+
+	ss.logger.Info(fmt.Sprintf("Nodepools in the cluster %v", nps))
+
+	if err != nil {
+		return err
+	}
+
+	var preemptibleNodePools, onDemandNodePools []container.NodePool
+
+	for _, np := range nps {
+		if np.Config.Preemptible {
+			preemptibleNodePools = append(preemptibleNodePools, *np)
+		} else {
+			onDemandNodePools = append(onDemandNodePools, *np)
+		}
+	}
+
+	// Loop over preemptible nodepools and find their corresponding fallback on-demand node-pool.
+	ss.nodePoolMap = make(map[string]npShiftConf)
+	for _, pnp := range preemptibleNodePools {
+		for _, onp := range onDemandNodePools {
+			if reflect.DeepEqual(pnp.Config.Labels, onp.Config.Labels) && pnp.Config.MachineType == onp.Config.MachineType {
```

Comment:
If Labels and MachineType is struct type consider using `go-cmp` instead of `reflect`

Reply:
Labels is a map and MachineType is a string. So reflect is enough here I believe.